spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/messaging/MessagingSpanInjector.java

```diff
@@ -29,24 +30,30 @@
 import org.springframework.util.StringUtils;
 
 /**
- * Utility for manipulating message headers related to span data.
+ * Creates a {@link Span.SpanBuilder} from {@link Message}
  *
- * @author Dave Syer
+ * @author Marcin Grzejszczak
  *
+ * @since 1.0.0
  */
-public class SpanMessageHeaders {
+class MessagingSpanInjector implements SpanInjector {
 
 	public static final String SPAN_HEADER = "X-Current-Span";
 
-	public static Span getSpanFromHeader(Message<?> message) {
-		if (message == null) {
-			return null;
-		}
-		Object object = message.getHeaders().get(SPAN_HEADER);
-		if (object instanceof Span) {
-			return (Span) object;
+	private final TraceKeys traceKeys;
+
+	public MessagingSpanInjector(TraceKeys traceKeys) {
+		this.traceKeys = traceKeys;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public <T> void inject(Span span, T carrier) {
```

Comment:
Cause:

- we want to be compatible with OpenTracing 
- it doesn't change much cause still I have to do sth like "accept" from Observator pattern